// convex/dist/cjs/server/impl/database_impl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupWriter = exports.setupReader = void 0;
const common_1 = require("@convex-dev/common");
const syscall_js_1 = require("./syscall.js");
const query_impl_js_1 = require("./query_impl.js");
function setupReader() {
    return {
        get: (id) => {
            const syscallJSON = (0, syscall_js_1.performSyscall)("get", { id: (0, common_1.convexToJson)(id) });
            const syscallResult = (0, common_1.jsonToConvex)(syscallJSON);
            return Promise.resolve(syscallResult);
        },
        table: (tableName) => new query_impl_js_1.TableImpl(tableName),
    };
}
exports.setupReader = setupReader;
function setupWriter() {
    const reader = setupReader();
    return {
        get: reader.get,
        table: reader.table,
        insert: (table, value) => {
            const syscallJSON = (0, syscall_js_1.performSyscall)("insert", {
                table,
                value: (0, common_1.convexToJson)(value),
            });
            const syscallResult = (0, common_1.jsonToConvex)(syscallJSON);
            return syscallResult._id;
        },
        patch: (id, value) => {
            (0, syscall_js_1.performSyscall)("update", {
                id: (0, common_1.convexToJson)(id),
                value: (0, common_1.convexToJson)(value),
            });
        },
        replace: (id, value) => {
            (0, syscall_js_1.performSyscall)("replace", {
                id: (0, common_1.convexToJson)(id),
                value: (0, common_1.convexToJson)(value),
            });
        },
        delete: id => {
            (0, syscall_js_1.performSyscall)("remove", { id: (0, common_1.convexToJson)(id) });
        },
    };
}
exports.setupWriter = setupWriter;
