// convex/dist/esm/server/impl/database_impl.js
import { convexToJson, jsonToConvex } from "@convex-dev/common";
import { performSyscall } from "./syscall.js";
import { TableImpl } from "./query_impl.js";
export function setupReader() {
    return {
        get: (id) => {
            const syscallJSON = performSyscall("get", { id: convexToJson(id) });
            const syscallResult = jsonToConvex(syscallJSON);
            return Promise.resolve(syscallResult);
        },
        table: (tableName) => new TableImpl(tableName),
    };
}
export function setupWriter() {
    const reader = setupReader();
    return {
        get: reader.get,
        table: reader.table,
        insert: (table, value) => {
            const syscallJSON = performSyscall("insert", {
                table,
                value: convexToJson(value),
            });
            const syscallResult = jsonToConvex(syscallJSON);
            return syscallResult._id;
        },
        patch: (id, value) => {
            performSyscall("update", {
                id: convexToJson(id),
                value: convexToJson(value),
            });
        },
        replace: (id, value) => {
            performSyscall("replace", {
                id: convexToJson(id),
                value: convexToJson(value),
            });
        },
        delete: id => {
            performSyscall("remove", { id: convexToJson(id) });
        },
    };
}
