// convex/dist/esm/server/impl/index_range_builder_impl.js
import { convexToJson } from "@convex-dev/common";
import { IndexRange, } from "../index_range_builder.js";
/**
 * @internal
 */
export class IndexRangeBuilderImpl extends IndexRange {
    constructor(range_expressions) {
        super();
        this.range_expressions = range_expressions;
    }
    static new() {
        return new IndexRangeBuilderImpl([]);
    }
    eq(fieldName, value) {
        return new IndexRangeBuilderImpl(this.range_expressions.concat({
            type: "Eq",
            fieldPath: fieldName,
            value: convexToJson(value),
        }));
    }
    gt(fieldName, value) {
        return new IndexRangeBuilderImpl(this.range_expressions.concat({
            type: "Gt",
            fieldPath: fieldName,
            value: convexToJson(value),
        }));
    }
    gte(fieldName, value) {
        return new IndexRangeBuilderImpl(this.range_expressions.concat({
            type: "Gte",
            fieldPath: fieldName,
            value: convexToJson(value),
        }));
    }
    lt(fieldName, value) {
        return new IndexRangeBuilderImpl(this.range_expressions.concat({
            type: "Lt",
            fieldPath: fieldName,
            value: convexToJson(value),
        }));
    }
    lte(fieldName, value) {
        return new IndexRangeBuilderImpl(this.range_expressions.concat({
            type: "Lte",
            fieldPath: fieldName,
            value: convexToJson(value),
        }));
    }
    export() {
        return this.range_expressions;
    }
}
