// convex/dist/esm/browser/sync/remote_query_set.js
import { jsonToConvex } from "@convex-dev/common";
import Long from "long";
import { logToConsole } from "../logging.js";
/**
 * A represention of the query results we've received on the current WebSocket
 * connection.
 */
export class RemoteQuerySet {
    constructor(queryPath) {
        this.version = { querySet: 0, ts: Long.fromNumber(0), identity: 0 };
        this.remoteQuerySet = new Map();
        this.queryPath = queryPath;
    }
    transition(transition) {
        const start = transition.startVersion;
        if (this.version.querySet !== start.querySet ||
            this.version.ts.notEquals(start.ts) ||
            this.version.identity !== start.identity) {
            throw new Error(`Invalid start version: ${start.ts}:${start.querySet}`);
        }
        for (const modification of transition.modifications) {
            switch (modification.type) {
                case "QueryUpdated": {
                    const queryPath = this.queryPath(modification.queryId);
                    if (queryPath) {
                        for (const line of modification.logLines) {
                            logToConsole("info", "query", queryPath, line);
                        }
                    }
                    const value = jsonToConvex(modification.value ?? null);
                    this.remoteQuerySet.set(modification.queryId, {
                        success: true,
                        value,
                    });
                    break;
                }
                case "QueryFailed": {
                    const queryPath = this.queryPath(modification.queryId);
                    if (queryPath) {
                        for (const line of modification.logLines) {
                            logToConsole("info", "query", queryPath, line);
                        }
                    }
                    this.remoteQuerySet.set(modification.queryId, {
                        success: false,
                        errorMessage: modification.errorMessage,
                    });
                    break;
                }
                case "QueryRemoved": {
                    this.remoteQuerySet.delete(modification.queryId);
                    break;
                }
                default: {
                    // Enforce that the switch-case is exhaustive.
                    // eslint-disable-next-line  @typescript-eslint/no-unused-vars
                    const _ = modification;
                    throw new Error(`Invalid modification ${modification}`);
                }
            }
        }
        this.version = transition.endVersion;
    }
    remoteQueryResults() {
        return this.remoteQuerySet;
    }
    timestamp() {
        return this.version.ts;
    }
}
